test/upgrade/prober/wathola/config/tracing.go

```diff
@@ -0,0 +1,33 @@
+/*
+Copyright 2022 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"go.uber.org/zap"
+	"knative.dev/pkg/tracing"
+	tracingconfig "knative.dev/pkg/tracing/config"
+)
+
+func SetupTracing() {
+	config, err := tracingconfig.JSONToTracingConfig(Instance.TracingConfig)
+	if err != nil {
+		Log.Warn("Tracing configuration is invalid, using the no-op default", zap.Error(err))
+	}
+	if err = tracing.SetupStaticPublishing(Log, "", config); err != nil {
```

Comment:
I'm thinking that maybe returning here the `*tracing.OpenCensusTracer` allows later to call [`Finish`](https://pkg.go.dev/knative.dev/pkg/tracing#OpenCensusTracer.Finish) method.

```go
func SetupStaticPublishing(logger *zap.SugaredLogger, serviceName string, cfg *config.Config) (*tracing.OpenCensusTracer, error) {
	oct := tracing.NewOpenCensusTracer(tracing.WithExporter(serviceName, logger))
	if err := oct.ApplyConfig(cfg); err != nil {
		return nil, fmt.Errorf("unable to set OpenCensusTracing config: %w", err)
	}
	return oct, nil
}
```

This method should call [`Close()`](https://github.com/knative/pkg/blob/1f7514acd0c2159b18cd1674c930298f0bbacc7a/vendor/github.com/openzipkin/zipkin-go/reporter/http/http.go#L71-L74) method of HTTP reporter, which should break the send loop, and try sending the last batch before shutdown. I think it should be enough to send all registered spans.

Reply:
Thanks. Let me try it.

Reply:
So this whole approach doesn't work because on [this line](https://github.com/knative/pkg/blob/1f7514acd0c2159b18cd1674c930298f0bbacc7a/tracing/opencensus.go#L92) there's nil passed to the function:  , and then on [this line](https://github.com/knative/pkg/blob/1f7514acd0c2159b18cd1674c930298f0bbacc7a/tracing/opencensus.go#L147) it throws "panic: runtime error: invalid memory address or nil pointer dereference"
It is also not possible to call tracing.SetupTracing() again (to force calling the reporter Close method) because the tracer is already registered and can't be done again.
As a side note, I've verified that 1.5 seconds is enough for the spans/traces to be sent properly.
I recommend using 1.5 seconds sleep time and call it done.